plotly/express/_core.py

```diff
@@ -1037,9 +1037,30 @@ def apply_default_cascade(args):
         if args["color_continuous_scale"] is None:
             args["color_continuous_scale"] = sequential.Viridis
 
+    # if color_discrete_sequence not set explicitly or in px.defaults,
+    # see if we can defer to template. Try trace-specific colors first,
+    # then layout.colorway, then set reasonable defaults
     if "color_discrete_sequence" in args:
+        if args["color_discrete_sequence"] is None and constructor is not None:
+            if constructor == "timeline":
+                trace_type = "bar"
+            else:
+                trace_type = constructor().type
+            if trace_data_list := getattr(args["template"].data, trace_type, None):
+                args["color_discrete_sequence"] = [
+                    trace_data.marker.color
+                    for trace_data in trace_data_list
+                    if hasattr(trace_data, "marker")
+                    and hasattr(trace_data.marker, "color")
+                ]
+                if not args["color_discrete_sequence"] or not any(
+                    args["color_discrete_sequence"]
+                ):
+                    args["color_discrete_sequence"] = None
```

Comment:
It would be a bit cleaner to store the colors in a temporary variable and then check if it's valid before assigning it to the args. 

```suggestion
                trace_specific_colors = [
                    trace_data.marker.color
                    for trace_data in trace_data_list
                    if hasattr(trace_data, "marker")
                    and hasattr(trace_data.marker, "color")
                ]
                # If template contains at least one color for this trace type, assign to color_discrete_sequence
                if trace_specific_colors and any(trace_specific_colors):
                    args["color_discrete_sequence"] = trace_specific_colors
```